Add example of GDExtension written pure C (#10403)
godotengine/godot-docs: tutorials/scripting/gdextension/gdextension_c_example.rst

Now, let's implement the functions we just declared. Create a file called
``init.c`` in the ``src`` folder and add this code:

.. code-block:: c

    #include "init.h"

    void initialize_gdexample_module(void *p_userdata, GDExtensionInitializationLevel p_level)
    {
    }

    void deinitialize_gdexample_module(void *p_userdata, GDExtensionInitializationLevel p_level)
    {
    }

    GDExtensionBool GDE_EXPORT gdexample_library_init(GDExtensionInterfaceGetProcAddress p_get_proc_address, GDExtensionClassLibraryPtr p_library, GDExtensionInitialization *r_initialization)
    {
        r_initialization->initialize = initialize_gdexample_module;
        r_initialization->deinitialize = deinitialize_gdexample_module;
        r_initialization->userdata = NULL;
        r_initialization->minimum_initialization_level = GDEXTENSION_INITIALIZATION_SCENE;

        return true;
    }

What this does is simply set up the initialization data that Godot expects. The
functions to initialize and deinitialize are set so Godot will call then when
needed. It also sets the initialization level which varies per extension. Since
we plan to add a custom node, the ``SCENE`` level is enough.

> We will fill the ``initialize_gdexample_module`` function later to register our custom class.

Author: Thanks! I found and fixed a couple additional ones

If you didn't compile the extension yet, it is the time to do it now. To do
that, open a terminal or command prompt, navigate to the root folder of the
extension and run ``scons``. It should compile quickly since the extension is
very simple.

Then, create a file called ``gdexample.gdextension`` inside the ``demo`` folder.
This is a Godot resource that describes the extension, allowing the engine to
properly load it. Put the following content in this file:

.. code-block::

    [configuration]

    entry_symbol = "gdexample_library_init"
    compatibility_minimum = "4.2"

    [libraries]
    macos.debug = "res://bin/libgdexample.dylib"
    linux.debug = "res://bin/libgdexample.so"
    windows.debug = "res://bin/libgdexample.dll"

As you can see, ``gdexample_library_init`` is the same name of the function we
defined in our ``init.c`` file. It is important that the names match because it
is how Godot calls the entry point of the extension.

We also set the compatibility minimum to 4.2, since we are targeting this
version. It should still work on later versions. If you are using a later Godot
version and rely on the new features, you need to increase this value to a

> version number that has everything you use.

Reviewer: Since this page will be added in version 4.4, it may be worth explaining a little bit more why 4.2 is chosen as the target. Maybe here or maybe also at the start of the page, where the requirement of a 4.2+ Godot engine version is mentioned.

But maybe that is already covered in previous GDExtension pages.

Author: Generally, GDExtensions will target the lowest version that will work, in order to avoid having to making lots of different builds for different Godot versions, because a GDExtension built for an older version should work in a newer version (where as one built for a newer version won't work in an older version).

I don't see anywhere in the docs where we already talk about that, but I'm not sure this is the right place to go into it?

Reviewer: Maybe another paragraph or sub-header under [Version Compatibility](https://docs.godotengine.org/en/latest/tutorials/scripting/gdextension/what_is_gdextension.html#version-compatibility), which can then be linked from other places as needed?

Author: Ah, that "Version Compatibility" section could certainly be improved! There's a warning further up on that same page which also say the thing about work on new versions but not older. Perhaps that warning should be removed and a similar statement put in "Version Compatibility"? Probably best in a separate PR, though

Reviewer: Yeah, that change makes sense to defer for a later PR.

It's also been a couple minor versions so maybe the compatibility policy is a bit more solid now?

Author: Here's a PR that attempts it: https://github.com/godotengine/godot-docs/pull/10404

Author: > It's also been a couple minor versions so maybe the compatibility policy is a bit more solid now?

Our messaging has always been a little overly paranoid. :-) Except for the break between 4.0 and 4.1, we've really maintained compatibility in 99.99% of cases. We probably should soften the language about this, but there's some debate about how exactly - see https://github.com/godotengine/godot-cpp/pull/1588

Reviewer: Yeah, that makes sense to me.

With https://github.com/godotengine/godot-docs/pull/10404 split out I think we can consider this resolved. If the other PR ends up merged before this one is merged, then maybe we should link to the explanation, but also maybe not since it appears in the intro page to GDExtension.

Reviewer: Seems fine.

It is _slightly_ strange that the opening mentions 4.2, then it takes half the article to explain why we use 4.2. If there is some natural way to fit this into the introduction, that would be nice. But I'm not sure if there is a natural way, and this is an advanced tutorial, so the current approach seems totally fine to me.

In the ``[libraries]`` section we set up the paths to the shared library on
different platforms. Here there's only the debug versions since that's what we

> are working on for the example. Using :ref:`Feature tags <doc_feature_tags>` you

Reviewer: If you're happy with the link showing the actual title of the page (in this case "Feature tags") you can omit the explicit link title:
```suggestion
are working on for the example. Using :ref:`doc_feature_tags` you
```
Alternately, you can use an explicit link title and make the link lowercase so its more natural in the context of a sentence:
```suggestion
are working on for the example. Using :ref:`feature tags <doc_feature_tags>` you
```

The current version seems worse than either option IMO

> can fine tune this to also provide release versions, add more target OSes, as

Reviewer: ```suggestion
can fine tune this to also provide release versions, add more target operating systems, as
```
"OSes" as a plural seems awkward. (And the docs have no existing usages of it)

well as providing 32-bit and 64-bit binaries.

You can also add library dependencies and custom icons for your classes in this
file, but this is out of the scope for this tutorial.

After saving the file, go back to the editor. Godot should automatically load

> the extension. Nothing will be seen because our extension only register a new

Reviewer: ```suggestion
the extension. Nothing will be seen because our extension only registers a new
```

> class. To use such, class add a ``Node2D`` as a root of the scene. Move it to

Reviewer: ```suggestion
class. To use such class add a ``Node2D`` as a root of the scene. Move it to
```
?

Alternately this seems more natural:
```suggestion
class. To use this class add a ``Node2D`` as a root of the scene. Move it to
```

the middle of viewport for better visibility. Then add a new child node to the

> root and in the ``Create New Node`` dialog search for "GDExample", the name of

Reviewer: ```suggestion
root and in the **Create New Node** dialog search for "GDExample", the name of
```
Bold for UI like dialogs

our class, as it should be listed there. If it isn't, it means that Godot didn't
load the extension properly, so try restarting the editor and retrace the steps
to see if anything went missing.

> Our custom class is derived from ``Sprite2D``, so it has a ``texture`` property

Reviewer: ```suggestion
Our custom class is derived from ``Sprite2D``, so it has a **Texture** property
```
Inspector properties are capitalized as the appear in the inspector (with the capitalization prettifying enabled) and bold, because they are editor UI.

(It's kinda inconsistently applied but it's the current "standard")

    {
        bind_method_0_r("GDExample", "get_amplitude", gdexample_class_get_amplitude, GDEXTENSION_VARIANT_TYPE_FLOAT);
        bind_method_1("GDExample", "set_amplitude", gdexample_class_set_amplitude, "amplitude", GDEXTENSION_VARIANT_TYPE_FLOAT);
        bind_property("GDExample", "amplitude", GDEXTENSION_VARIANT_TYPE_FLOAT, "get_amplitude", "set_amplitude");

        bind_method_0_r("GDExample", "get_speed", gdexample_class_get_speed, GDEXTENSION_VARIANT_TYPE_FLOAT);
        bind_method_1("GDExample", "set_speed", gdexample_class_set_speed, "speed", GDEXTENSION_VARIANT_TYPE_FLOAT);
        bind_property("GDExample", "speed", GDEXTENSION_VARIANT_TYPE_FLOAT, "get_speed", "set_speed");
    }

If you build the extension with ``scons``, you'll see in the Godot editor the new property shown
not only on the documentation page for the custom class but also in the Inspector dock when the
``GDExample`` node is selected.

.. image:: img/gdextension_c_inspector_properties.webp

Binding virtual methods
-----------------------

Our custom node now has properties to influence how it operates, but it still
doesn't do anything. In this section, we will bind the virtual method

> :ref:`_process<class_Node_private_method__process>` and make our custom sprite

Reviewer: ```suggestion
:ref:`_process() <class_Node_private_method__process>` and make our custom sprite
```

move a little bit.

In the ``gdexample.h`` file, let's add a function that represents the custom

> ``_process`` method:

Reviewer: ```suggestion
``_process()`` method:
```

        ...
        GDExtensionInterfaceVariantGetPtrOperatorEvaluator variant_get_ptr_operator_evaluator = (GDExtensionInterfaceVariantGetPtrOperatorEvaluator)p_get_proc_address("variant_get_ptr_operator_evaluator");

        ...

        // Operators.
        operators.string_name_equal = variant_get_ptr_operator_evaluator(GDEXTENSION_VARIANT_OP_EQUAL, GDEXTENSION_VARIANT_TYPE_STRING_NAME, GDEXTENSION_VARIANT_TYPE_STRING_NAME);
    }

As you can see we need a new local helper here in order to grab the function
pointer for the operator.

With this handy, we can easily create our comparison function in the same file:

.. code-block:: c

    bool is_string_name_equal(GDExtensionConstStringNamePtr p_a, const char *p_b)
    {
        // Create a StringName for the C string.
        StringName string_name;
        constructors.string_name_new_with_latin1_chars(&string_name, p_b);

        // Compare both StringNames.
        bool is_equal = false;
        operators.string_name_equal(p_a, &string_name, &is_equal);

        // Destroy the created StringName.
        destructors.string_name_destructor(&string_name);

        // Return the result.
        return is_equal;
    }

> This function simply creates a ``StringName`` from the argument, compares with

Reviewer: ```suggestion
This function creates a ``StringName`` from the argument, compares with
```

        if (is_string_name_equal(p_name, "_process"))
        {
            return (void *)gdexample_class_process;
        }
        // Otherwise, return NULL.
        return NULL;
    }

    void gdexample_call_virtual_with_data(GDExtensionClassInstancePtr p_instance, GDExtensionConstStringNamePtr p_name, void *p_virtual_call_userdata, const GDExtensionConstTypePtr *p_args, GDExtensionTypePtr r_ret)
    {
        // If it is the "_process" method, call it with a helper.
        if (p_virtual_call_userdata == &gdexample_class_process)
        {
            ptrcall_1_float_arg_no_ret(p_virtual_call_userdata, p_instance, p_args, r_ret);
        }
    }

Those functions are also quite straightforward after making all the helpers
previously.

For the first one, we simply check if the function name requested is
``_process`` and if it is we return a function pointer to our implementation of
it. Otherwise we return ``NULL``, signaling that the method is not being
overridden. We don't use the ``p_class_userdata`` here since this function is
meant only for one class and we don't have any data associated with it.

> The second one is similar. If it is the ``_process`` method, it uses the given

Reviewer: ```suggestion
The second one is similar. If it is the ``_process()`` method, it uses the given
```

function pointer to call the ``ptrcall`` helper, passing the call arguments
forward. Otherwise it simply does nothing, since we don't have any other virtual
methods being implemented.

The only thing missing is using those callbacks when the class is registered. Go
to the ``init.c`` file and change the ``class_info`` initialization to include
those, replacing the ``NULL`` value used previously:

.. code-block:: c

    void initialize_gdexample_module(void *p_userdata, GDExtensionInitializationLevel p_level)
    {
        ...

        GDExtensionClassCreationInfo2 class_info = {
            ...
            .get_virtual_call_data_func = gdexample_get_virtual_with_data,
            .call_virtual_with_data_func = gdexample_call_virtual_with_data,
            ...
        };

        ...
    }

This is enough to bind the virtual method. If you build the extension and run

> the demo project again, the ``_process`` function will be called. You just won't

Reviewer: ```suggestion
the demo project again, the ``_process()`` function will be called. You just won't
```

be able to tell since the function itself does nothing visible. We will solve
this now by making the custom node move following a pattern.

In order to make our node do stuff, we'll need to call Godot methods. Not only
the GDExtension API functions as we've being doing so far, but actual engine
methods, as we would do with scripting. This naturally requires some extra setup.

> First, let's add :ref:`Vector2 <class_Vector2>` to our ``defs.h`` file, so we

Reviewer: ```suggestion
First, let's add :ref:`class_Vector2` to our ``defs.h`` file, so we
```
The explicit link title *can* be omitted here but in many existing cases it is not. So this suggestion is optional.

    {
        self->time_passed += self->speed * delta;

        Vector2 new_position;

        // Set up the arguments for the Vector2 constructor.
        double x = self->amplitude + (self->amplitude * sin(self->time_passed * 2.0));
        double y = self->amplitude + (self->amplitude * cos(self->time_passed * 1.5));
        GDExtensionConstTypePtr args[] = {&x, &y};
        // Call the Vector2 constructor.
        constructors.vector2_constructor_x_y(&new_position, args);

        // Set up the arguments for the set_position method.
        GDExtensionConstTypePtr args2[] = {&new_position};
        // Call the set_position method.
        api.object_method_bind_ptrcall(methods.node2d_set_position, self->object, args2, NULL);
    }

After updating the time passed scaled by the ``speed`` property, it creates
``x`` and ``y`` values based on that, also modulated by the ``amplitude``
property. This is what will give the pattern effect. The ``math.h`` header is
needed for the ``sin()`` and ``cos()`` functions used here.

Then it sets up an array of arguments to construct a ``Vector2``, followed by
calling the constructor. It sets up another array of arguments and use it to

> call the ``set_position`` method via the bind we acquired previously.

Reviewer: ```suggestion
call the ``set_position()`` method via the bind we acquired previously.
```

Since nothing here allocates any memory, there's not a need to cleanup.

Now we can build the extension again and reopen Godot. Even in the editor you'll
see the custom sprite moving.

.. image:: img/gdextension_c_moving_sprite.gif

> Try changing the ``Speed`` and ``Amplitude`` properties and see how the sprite

Reviewer: ```suggestion
Try changing the **Speed** and **Amplitude** properties and see how the sprite
```
Like this if these are in the inspector.

        StringName signal_string_name;
        constructors.string_name_new_with_latin1_chars(&signal_string_name, signal_name);

        GDExtensionPropertyInfo args_info[] = {
            make_property(arg1_type, arg1_name),
        };

        api.classdb_register_extension_class_signal(class_library, &class_string_name, &signal_string_name, args_info, 1);

        // Destruct things.
        destructors.string_name_destructor(&class_string_name);
        destructors.string_name_destructor(&signal_string_name);
        destruct_property(&args_info[0]);
    }

This one is very similar to the function to bind methods. The main difference is
that we don't need to fill another struct, we just pass the needed names and the
array of arguments. The ``1`` at the end means the amount of arguments the
signal provides.

With this we can bind the signal in ``gdexample.c``:

.. code-block:: c

    void gdexample_class_bind_methods()
    {
        ...
        bind_signal_1("GDExample", "position_changed", "new_position", GDEXTENSION_VARIANT_TYPE_VECTOR2);
    }

In order to emit a signal, we need to call the

> :ref:`emit_signal<class_Object_method_emit_signal>` method on our custom node.

Reviewer: ```suggestion
:ref:`emit_signal() <class_Object_method_emit_signal>` method on our custom node.
```

    }

Now we can go back to ``gdexample.c`` and define the new functions, without forgetting to
include the ``api.h`` header:

.. code-block:: c

    #include "gdexample.h"

    #include "api.h"

    ...

    const GDExtensionInstanceBindingCallbacks gdexample_class_binding_callbacks = {
        .create_callback = NULL,
        .free_callback = NULL,
        .reference_callback = NULL,
    };

    GDExtensionObjectPtr gdexample_create_instance(void *p_class_userdata)
    {
        // Create native Godot object;
        StringName class_name;
        constructors.string_name_new_with_latin1_chars(&class_name, "Sprite2D");
        GDExtensionObjectPtr object = api.classdb_construct_object(&class_name);
        destructors.string_name_destructor(&class_name);

        // Create extension object.
        GDExample *self = (GDExample *)api.mem_alloc(sizeof(GDExample));
        gdexample_class_constructor(self);
        self->object = object;

        // Set the extension instance in the native Godot object.
        constructors.string_name_new_with_latin1_chars(&class_name, "GDExample");
        api.object_set_instance(object, &class_name, self);

>         api.object_set_instance_binding(object, class_library, &gdexample_class_binding_callbacks);

Author: Thanks!

Sorry, I had intended to test all the code changes by updating vnen's complete repo, but hadn't gotten around to it yet. This one is now updated, plus a few others :-)
